[SPARK-17360][PYSPARK] Support generator in createDataFrame

gsemet opened the pull request:
## What changes were proposed in this pull request?

Avoid useless iteration within 'data' structure when creating a data frame from local data (so when createDataFrame uses _createFromLocal) and when 'data' is a Python generator (instead of a 'list' or a dataframe)
- if schema is provided, only iterate once (in _createFromLocal). In this case we skip 1 struction iteration
- if no schema is provided, we skip the useless 'map' with the 'identify' lambda function

There are still serialization to the JVM.
## How was this patch tested?

Pyspark unit tests and standalone manual pyspark execution

SparkQA commented:
**[Test build #64778 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64778/consoleFull)** for PR 14918 at commit [`59a6470`](https://github.com/apache/spark/commit/59a64707745f44465f82b326843ba115c1414d7a).

SparkQA commented:
**[Test build #64778 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64778/consoleFull)** for PR 14918 at commit [`59a6470`](https://github.com/apache/spark/commit/59a64707745f44465f82b326843ba115c1414d7a).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64778/
Test FAILed.

Stibbons commented on python/pyspark/sql/session.py:
```diff
@@ -396,14 +398,18 @@ def _createFromLocal(self, data, schema):
+        if prepareFunc is not None:
+            data = [schema.toInternal(prepareFunc(row)) for row in data]
+        else:
+            data = [schema.toInternal(row) for row in data]
```
in this case we avoided the call to ```map(prepare, data)``` when prepare is the identity function

SparkQA commented:
**[Test build #64782 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64782/consoleFull)** for PR 14918 at commit [`3353a46`](https://github.com/apache/spark/commit/3353a46c214d7917cc0fe659daabbe76ae9521e6).

Stibbons commented on python/pyspark/sql/session.py:
```diff
@@ -373,16 +375,16 @@ def _createFromRDD(self, rdd, schema, samplingRatio):
 
         if schema is None or isinstance(schema, (list, tuple)):
+            # make sure data could consumed multiple times
+            if not isinstance(data, list):
```
here we can convert to list because there will be several iterations on the data

Stibbons commented on python/pyspark/sql/session.py:
```diff
@@ -396,14 +398,18 @@ def _createFromLocal(self, data, schema):
         # convert python objects to sql data
-        data = [schema.toInternal(row) for row in data]
+        if prepareFunc is not None:
+            data = [schema.toInternal(prepareFunc(row)) for row in data]
```
by executing prepare at the same time than 'toInternal', we avoided 1 loop

Stibbons commented on python/pyspark/sql/tests.py:
```diff
@@ -196,7 +199,8 @@ def setUpClass(cls):
         os.unlink(cls.tempdir.name)
         cls.spark = SparkSession(cls.sc)
-        cls.testData = [Row(key=i, value=str(i)) for i in range(100)]
+        cls.testData = (Row(key=i, value=str(i)) for i in xrange(100))
```
change the unit test to use generator by default

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64782/
Test FAILed.

SparkQA commented:
**[Test build #64939 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64939/consoleFull)** for PR 14918 at commit [`e80170f`](https://github.com/apache/spark/commit/e80170f7a11c2037aa2b77e1a80443f25264d3bf).

SparkQA commented:
**[Test build #64939 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64939/consoleFull)** for PR 14918 at commit [`e80170f`](https://github.com/apache/spark/commit/e80170f7a11c2037aa2b77e1a80443f25264d3bf).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64939/
Test FAILed.

SparkQA commented:
**[Test build #64946 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64946/consoleFull)** for PR 14918 at commit [`61cdabf`](https://github.com/apache/spark/commit/61cdabf4c0d7f5cd3c31946993f6520a9cb26062).

SparkQA commented:
**[Test build #64946 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64946/consoleFull)** for PR 14918 at commit [`61cdabf`](https://github.com/apache/spark/commit/61cdabf4c0d7f5cd3c31946993f6520a9cb26062).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64946/
Test FAILed.

SparkQA commented:
**[Test build #64947 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64947/consoleFull)** for PR 14918 at commit [`5ff0bf4`](https://github.com/apache/spark/commit/5ff0bf46323606bc871e9d5a490f0be69ce1453e).

SparkQA commented:
**[Test build #64947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64947/consoleFull)** for PR 14918 at commit [`5ff0bf4`](https://github.com/apache/spark/commit/5ff0bf46323606bc871e9d5a490f0be69ce1453e).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64947/
Test FAILed.

SparkQA commented:
**[Test build #64948 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64948/consoleFull)** for PR 14918 at commit [`3e7a003`](https://github.com/apache/spark/commit/3e7a0030b8360892b70eebfb5af0ebd09f9b5591).

SparkQA commented:
**[Test build #64948 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64948/consoleFull)** for PR 14918 at commit [`3e7a003`](https://github.com/apache/spark/commit/3e7a0030b8360892b70eebfb5af0ebd09f9b5591).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64948/
Test FAILed.

SparkQA commented:
**[Test build #64951 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64951/consoleFull)** for PR 14918 at commit [`6f28f75`](https://github.com/apache/spark/commit/6f28f75c455b25b9a9faeec15b32b9bbe8fc8a5a).

SparkQA commented:
**[Test build #64951 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64951/consoleFull)** for PR 14918 at commit [`6f28f75`](https://github.com/apache/spark/commit/6f28f75c455b25b9a9faeec15b32b9bbe8fc8a5a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64951/
Test FAILed.

SparkQA commented:
**[Test build #64985 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64985/consoleFull)** for PR 14918 at commit [`45eadb0`](https://github.com/apache/spark/commit/45eadb05f1930d7e1248bb23044d4d64db57028d).

SparkQA commented:
**[Test build #64985 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64985/consoleFull)** for PR 14918 at commit [`45eadb0`](https://github.com/apache/spark/commit/45eadb05f1930d7e1248bb23044d4d64db57028d).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64985/
Test FAILed.

SparkQA commented:
**[Test build #64989 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64989/consoleFull)** for PR 14918 at commit [`0db5883`](https://github.com/apache/spark/commit/0db588331d3b73ddc326dfbbdb7ba3f39d5a7637).

SparkQA commented:
**[Test build #64989 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64989/consoleFull)** for PR 14918 at commit [`0db5883`](https://github.com/apache/spark/commit/0db588331d3b73ddc326dfbbdb7ba3f39d5a7637).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64989/
Test FAILed.

SparkQA commented:
**[Test build #64994 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64994/consoleFull)** for PR 14918 at commit [`639d149`](https://github.com/apache/spark/commit/639d1491a740fb6b32186aeb3937cd74f5783b33).

SparkQA commented:
**[Test build #64994 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64994/consoleFull)** for PR 14918 at commit [`639d149`](https://github.com/apache/spark/commit/639d1491a740fb6b32186aeb3937cd74f5783b33).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64994/
Test FAILed.

gsemet committed:
[SPARK-17360][PYSPARK] Support generator in createDataFrame

Avoid useless iteration within 'data' structure if it is
a Python generator, instead of dict or list
(so when createDataFrame uses _createFromLocal)

- if schema is provided, only iterate once (in _createFromLocal)
  In this case we skip 1 iteration
- if no schema is provided, we skip the useless 'map' with
  the identify function

There are still serialization to the JVM.

Signed-off-by: Gaetan Semet <gaetan@xeberon.net>

SparkQA commented:
**[Test build #64995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64995/consoleFull)** for PR 14918 at commit [`fa4c66c`](https://github.com/apache/spark/commit/fa4c66c03946cadb6f074a8d9ac12ec923e0ec18).

SparkQA commented:
**[Test build #64995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/64995/consoleFull)** for PR 14918 at commit [`fa4c66c`](https://github.com/apache/spark/commit/fa4c66c03946cadb6f074a8d9ac12ec923e0ec18).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/64995/
Test PASSed.

Stibbons commented:
at last, tests pass ! :)

holdenk commented:
So I'm not super sure I understand the point of the refactor - at some point we are forced to materialize the entire contents of the generator on the driver anyways to copy to the JVM since parallelize only has a special case for xrange not all generators. (Although I may be missing something here).

Stibbons commented:
Indeed, i dont think it will be feasible to propagate the generator up to the jvm. It would be cool, because when we have the schema there is no need to iterate several time on the complete sequence. But, at least with this patch we can prevent one (no schema) or two (when schema is provided) useless iterations of the whole sequence using a generator + this optim, it is still a good thing (Python is no so good sadly). I can try to do some performance evaluation, not before next week.

Stibbons commented:
Rebased

SparkQA commented:
**[Test build #65761 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65761/consoleFull)** for PR 14918 at commit [`01a9580`](https://github.com/apache/spark/commit/01a958073bb33dc8f9a536696eb3440b7b24fd02).

SparkQA commented:
**[Test build #65761 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65761/consoleFull)** for PR 14918 at commit [`01a9580`](https://github.com/apache/spark/commit/01a958073bb33dc8f9a536696eb3440b7b24fd02).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65761/
Test PASSed.

Stibbons commented:
For information, I continue to look at these kind of simple optimisations that does not cost too much. Python is a pretty slow language, very productive in term of code writing, but inefficient in term of execution. So if we can avoid looping once or twice in a list and replace lists by generator, I think that is very valuable for the Spark project

SparkQA commented:
**[Test build #66042 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66042/consoleFull)** for PR 14918 at commit [`76ccc50`](https://github.com/apache/spark/commit/76ccc50de470946295189351d778b20333e8b616).

SparkQA commented:
**[Test build #66042 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66042/consoleFull)** for PR 14918 at commit [`76ccc50`](https://github.com/apache/spark/commit/76ccc50de470946295189351d778b20333e8b616).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66042/
Test PASSed.

holdenk commented:
I'm not sure about its value - since it only applies to equivalent of parallelize which by definition should be used on relatively small datasets.

SparkQA commented:
**[Test build #66388 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66388/consoleFull)** for PR 14918 at commit [`8b85d63`](https://github.com/apache/spark/commit/8b85d63e0498def276bdcb945986ce3776ec6531).

SparkQA commented:
**[Test build #66388 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66388/consoleFull)** for PR 14918 at commit [`8b85d63`](https://github.com/apache/spark/commit/8b85d63e0498def276bdcb945986ce3776ec6531).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66388/
Test PASSed.

holdenk commented:
Do we have any usecases or benchmarks for cases where this would be helpful?

HyukjinKwon commented:
I do not agree with this change too by the same reason in https://github.com/apache/spark/pull/14918#issuecomment-250882422.

Stibbons commented:
Ok, abandonning

xor007 commented:
> Do we have any usecases or benchmarks for cases where this would be helpful?

Yes my huge use case which I am surprised a lot of people in industry don't have is **massive data mining**:

- You have a lot of files on the internet (for instance text from a large collection of webpages)
- You are able to write a python generator that goes through the files to find and ouput sentences containing the word "covfefe": I have seen a python generator go through 90G of such a real collection of 11000 files within minutes(they where downloaded)
- You want to create a dataframe of all those sentences and the actual collection of those sentences ends up being less than 20Mb

You could create a Dataset from the generator. Now that I have written this it seems I can run flatmap on the file with the generator as the transformation.

But something like Dataframe.from_generator in spark would be nice.

HyukjinKwon commented:
You can already do all the thins above via other APIs like `rdd.mapPartition`.

xor007 commented:
thanks @HyukjinKwon ! will look into this.
